nexus-repository-r-it/src/test/java/org/sonatype/nexus/plugins/r/internal/RHostedIT.java

```diff
@@ -0,0 +1,131 @@
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+
+import org.sonatype.nexus.common.app.BaseUrlHolder;
+import org.sonatype.nexus.pax.exam.NexusPaxExamSupport;
+import org.sonatype.nexus.plugins.r.internal.fixtures.RepositoryRuleR;
+import org.sonatype.nexus.repository.Repository;
+import org.sonatype.nexus.repository.storage.Asset;
+import org.sonatype.nexus.repository.storage.Component;
+import org.sonatype.nexus.repository.storage.ComponentMaintenance;
+import org.sonatype.nexus.testsuite.testsupport.NexusITSupport;
+
+import org.apache.commons.compress.compressors.CompressorStreamFactory;
+import org.apache.http.HttpResponse;
+import org.apache.http.entity.ByteArrayEntity;
+import org.apache.tika.io.IOUtils;
+import org.hamcrest.Matcher;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.ops4j.pax.exam.Configuration;
+import org.ops4j.pax.exam.Option;
+
+import static org.apache.commons.compress.compressors.CompressorStreamFactory.GZIP;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.isEmptyString;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.core.IsEqual.equalTo;
+
+public class RHostedIT
+    extends RITSupport
+{
+  private RClient client;
+
+  private Repository repository;
+
+  @Rule
+  public RepositoryRuleR repos = new RepositoryRuleR(() -> repositoryManager);
+
+  @Configuration
+  public static Option[] configureNexus() {
+    return NexusPaxExamSupport.options(
+        NexusITSupport.configureNexusBase(),
+        nexusFeature("org.sonatype.nexus.plugins", "nexus-repository-r")
+    );
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    BaseUrlHolder.set(this.nexusUrl.toString());
+    repository = repos.createRHosted("r-hosted-test");
+    client = createRHostedClient(repository);
+  }
+
+  @Test
+  public void testPackagesProcessing() throws Exception
+  {
+    assertThat(getAllComponents(repository), hasSize(0));
+    final File file = testData.resolveFile(AGRICOLAE_PKG_FILE_NAME_131_TGZ);
+    client.put(AGRICOLAE_PATH_FULL_131_TGZ, new ByteArrayEntity(Files.readAllBytes(Paths.get(file.getAbsolutePath()))));
+
+    //Verify DB contains data about uploaded component and asset
+    Component component = findComponent(repository, AGRICOLAE_PKG_NAME);
+    assertThat(component.name(), is(equalTo(AGRICOLAE_PKG_NAME)));
+    assertThat(component.version(), is(equalTo(AGRICOLAE_PKG_VERSION_131)));
+
+    //Verify Asset is created.
+    Asset asset = findAsset(repository, AGRICOLAE_PATH_FULL_131_TGZ);
+    assertThat(asset.name(), is(equalTo(AGRICOLAE_PATH_FULL_131_TGZ)));
+    assertThat(asset.format(), is(equalTo(R_FORMAT_NAME)));
+
+    //Verify that package could be downloaded from NXRM
+    HttpResponse resp = client.fetch(AGRICOLAE_PATH_FULL_131_TGZ);
```

Comment:
Suggestion: Make this a separate test, wrap the test above into a method that can be reused here for upload

Reply:
ec3e07bd99b71ebdd951f01522ad4fd1a3ff0373